### ISSUE TITLE
Unable to add new block to lwc_blocks!

### ISSUE THREAD

**Issue by gorogoro-space** [pop4959/LWCX#43]
**LWC Version (/lwc admin version):** 
```
LWC
http://griefcraft.com
Main plugin: 2.1.5-09ad392/
```

**Server version (/version):** 
```
This server is running CraftBukkit version git-Spigot-1a3504a-e7700ab (MC: 1.13.2)
 (Implementing API version 1.13.2-R0.1-SNAPSHOT)
This is a final build for 1.13.2. Please see https://www.spigotmc.org/go/1.13.2 for details about upgrading.
```

**Short five word summary of issue:** 
Right-clicking on the sign displays an error and does not protect.

**Detailed description of problem:**
Right-clicking on the sign displays an error and does not protect.
An error occurs at the first right-click.
It also occurs when a new sign is posted.
No error will occur after the first time. The reproducibility is 100%.

**If there was an error in console, put it here:**
```
[01:56:52] [Server thread/INFO]: [LWC] Unable to add new block to lwc_blocks!
[01:56:52] [Server thread/WARN]: org.sqlite.SQLiteException: [SQLITE_CONSTRAINT]  Abort due to constraint violation (UNIQUE constraint failed: lwc_blocks.id)
[01:56:52] [Server thread/WARN]: 	at org.sqlite.core.DB.newSQLException(DB.java:941)
[01:56:52] [Server thread/WARN]: 	at org.sqlite.core.DB.newSQLException(DB.java:953)
[01:56:52] [Server thread/WARN]: 	at org.sqlite.core.DB.execute(DB.java:854)
[01:56:52] [Server thread/WARN]: 	at org.sqlite.core.DB.executeUpdate(DB.java:895)
[01:56:52] [Server thread/WARN]: 	at org.sqlite.jdbc3.JDBC3PreparedStatement.executeUpdate(JDBC3PreparedStatement.java:102)
[01:56:52] [Server thread/WARN]: 	at com.griefcraft.cache.BlockCache.addBlock(BlockCache.java:150)
[01:56:52] [Server thread/WARN]: 	at com.griefcraft.cache.BlockCache.getBlockId(BlockCache.java:253)
[01:56:52] [Server thread/WARN]: 	at com.griefcraft.cache.BlockCache.getBlockId(BlockCache.java:277)
[01:56:52] [Server thread/WARN]: 	at com.griefcraft.model.Protection.isBlockInWorld(Protection.java:628)
[01:56:52] [Server thread/WARN]: 	at com.griefcraft.util.ProtectionFinder.tryLoadProtection(ProtectionFinder.java:309)
[01:56:52] [Server thread/WARN]: 	at com.griefcraft.util.ProtectionFinder.matchBlocks(ProtectionFinder.java:117)
[01:56:52] [Server thread/WARN]: 	at com.griefcraft.lwc.LWC.findProtection(LWC.java:1333)
[01:56:52] [Server thread/WARN]: 	at com.griefcraft.lwc.LWC.findProtection(LWC.java:1306)
[01:56:52] [Server thread/WARN]: 	at com.griefcraft.lwc.LWC.findProtection(LWC.java:1296)
[01:56:52] [Server thread/WARN]: 	at com.griefcraft.listeners.LWCPlayerListener.onPlayerInteract(LWCPlayerListener.java:624)
[01:56:52] [Server thread/WARN]: 	at sun.reflect.GeneratedMethodAccessor1372.invoke(Unknown Source)
[01:56:52] [Server thread/WARN]: 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[01:56:52] [Server thread/WARN]: 	at java.lang.reflect.Method.invoke(Method.java:498)
[01:56:52] [Server thread/WARN]: 	at org.bukkit.plugin.java.JavaPluginLoader$1.execute(JavaPluginLoader.java:312)
[01:56:52] [Server thread/WARN]: 	at org.bukkit.plugin.RegisteredListener.callEvent(RegisteredListener.java:66)
[01:56:52] [Server thread/WARN]: 	at org.bukkit.plugin.SimplePluginManager.fireEvent(SimplePluginManager.java:507)
[01:56:52] [Server thread/WARN]: 	at org.bukkit.plugin.SimplePluginManager.callEvent(SimplePluginManager.java:492)
[01:56:52] [Server thread/WARN]: 	at org.bukkit.craftbukkit.v1_13_R2.event.CraftEventFactory.callPlayerInteractEvent(CraftEventFactory.java:304)
[01:56:52] [Server thread/WARN]: 	at net.minecraft.server.v1_13_R2.PlayerInteractManager.a(PlayerInteractManager.java:458)
[01:56:52] [Server thread/WARN]: 	at net.minecraft.server.v1_13_R2.PlayerConnection.a(PlayerConnection.java:1222)
[01:56:52] [Server thread/WARN]: 	at net.minecraft.server.v1_13_R2.PacketPlayInUseItem.a(PacketPlayInUseItem.java:37)
[01:56:52] [Server thread/WARN]: 	at net.minecraft.server.v1_13_R2.PacketPlayInUseItem.a(PacketPlayInUseItem.java:1)
[01:56:52] [Server thread/WARN]: 	at net.minecraft.server.v1_13_R2.PlayerConnectionUtils.lambda$0(PlayerConnectionUtils.java:9)
[01:56:52] [Server thread/WARN]: 	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
[01:56:52] [Server thread/WARN]: 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
[01:56:52] [Server thread/WARN]: 	at net.minecraft.server.v1_13_R2.SystemUtils.a(SourceFile:199)
[01:56:52] [Server thread/WARN]: 	at net.minecraft.server.v1_13_R2.MinecraftServer.b(MinecraftServer.java:896)
[01:56:52] [Server thread/WARN]: 	at net.minecraft.server.v1_13_R2.DedicatedServer.b(DedicatedServer.java:417)
[01:56:52] [Server thread/WARN]: 	at net.minecraft.server.v1_13_R2.MinecraftServer.a(MinecraftServer.java:831)
[01:56:52] [Server thread/WARN]: 	at net.minecraft.server.v1_13_R2.MinecraftServer.run(MinecraftServer.java:729)
[01:56:52] [Server thread/WARN]: 	at java.lang.Thread.run(Thread.java:748)
```

**Comment by pop4959** [comment]
Unable to reproduce the error. Can you try updating to the latest version of the plugin [here](https://www.spigotmc.org/resources/lwc-extended.69551/)? It looks like yours is out of date. If you are not aware, LWCX replaces Modern LWC, which I am no longer updating. There were a few bugs with signs that got fixed after 2.1.5 was released.

Let me know if that fixes your issue, otherwise we may have to look a bit deeper. It may be a problem with your specific database. Try updating first though. :smiley:

**Comment by gorogoro-space** [comment]
I replaced Modern LWC with LWCX and restarted.
Error reproducibility is low.
`/lwc info` shows that it should has been protected, but it was not protected.

It may be a problem with my specific database.

# DATABASE
```
# sqlite3 lwc.db
SQLite version 3.X.XX YYYY-MM-DD HH:MI:SS
Enter ".help" for instructions
Enter SQL statements terminated with a ";"
sqlite> .tables
lwc_blocks       lwc_history      lwc_internal     lwc_protections
sqlite> .schema lwc_blocks
CREATE TABLE lwc_blocks ( id INTEGER PRIMARY KEY , name VARCHAR(40)  );
sqlite> select * from lwc_blocks;
0|AIR
23|DISPENSER
54|CHEST
61|FURNACE
62|FURNACE
63|LEGACY_SIGN
64|OAK_DOOR
68|LEGACY_WALL_SIGN
71|IRON_DOOR
96|OAK_TRAPDOOR
107|OAK_FENCE_GATE
146|TRAPPED_CHEST
183|SPRUCE_FENCE_GATE
184|BIRCH_FENCE_GATE
185|JUNGLE_FENCE_GATE
186|DARK_OAK_FENCE_GATE
193|SPRUCE_DOOR
194|BIRCH_DOOR
195|JUNGLE_DOOR
196|ACACIA_DOOR
197|DARK_OAK_DOOR
198|OAK_WALL_SIGN
sqlite>
```

**LWC Version (/lwc admin version):** 
```
LWC
http://griefcraft.com
Main plugin: 2.2.1-5272a54/
```

**Server version (/version):** 
```
This server is running CraftBukkit version git-Spigot-1a3504a-e7700ab (MC: 1.13.2)
 (Implementing API version 1.13.2-R0.1-SNAPSHOT)
This is a final build for 1.13.2. Please see https://www.spigotmc.org/go/1.13.2 for details about upgrading.
```

**If there was an error in console, put it here:**
```
[06:41:26] [Server thread/INFO]: [LWC] Unable to add new block to lwc_blocks!
[06:41:26] [Server thread/WARN]: org.sqlite.SQLiteException: [SQLITE_CONSTRAINT]  Abort due to constraint violation (UNIQUE constraint failed: lwc_blocks.id)
[06:41:26] [Server thread/WARN]:        at org.sqlite.core.DB.newSQLException(DB.java:941)
[06:41:26] [Server thread/WARN]:        at org.sqlite.core.DB.newSQLException(DB.java:953)
[06:41:26] [Server thread/WARN]:        at org.sqlite.core.DB.execute(DB.java:854)
[06:41:26] [Server thread/WARN]:        at org.sqlite.core.DB.executeUpdate(DB.java:895)
[06:41:26] [Server thread/WARN]:        at org.sqlite.jdbc3.JDBC3PreparedStatement.executeUpdate(JDBC3PreparedStatement.java:102)
[06:41:26] [Server thread/WARN]:        at com.griefcraft.cache.BlockCache.addBlock(BlockCache.java:150)
[06:41:26] [Server thread/WARN]:        at com.griefcraft.cache.BlockCache.getBlockId(BlockCache.java:253)
[06:41:26] [Server thread/WARN]:        at com.griefcraft.cache.BlockCache.getBlockId(BlockCache.java:277)
[06:41:26] [Server thread/WARN]:        at com.griefcraft.model.Protection.isBlockInWorld(Protection.java:634)
[06:41:26] [Server thread/WARN]:        at com.griefcraft.util.ProtectionFinder.tryLoadProtection(ProtectionFinder.java:309)
[06:41:26] [Server thread/WARN]:        at com.griefcraft.util.ProtectionFinder.matchBlocks(ProtectionFinder.java:117)
[06:41:26] [Server thread/WARN]:        at com.griefcraft.lwc.LWC.findProtection(LWC.java:1372)
[06:41:26] [Server thread/WARN]:        at com.griefcraft.lwc.LWC.findProtection(LWC.java:1345)
[06:41:26] [Server thread/WARN]:        at com.griefcraft.lwc.LWC.findProtection(LWC.java:1335)
[06:41:26] [Server thread/WARN]:        at com.griefcraft.listeners.LWCPlayerListener.onPlayerInteract(LWCPlayerListener.java:630)
[06:41:26] [Server thread/WARN]:        at sun.reflect.GeneratedMethodAccessor125.invoke(Unknown Source)
[06:41:26] [Server thread/WARN]:        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[06:41:26] [Server thread/WARN]:        at java.lang.reflect.Method.invoke(Method.java:498)
[06:41:26] [Server thread/WARN]:        at org.bukkit.plugin.java.JavaPluginLoader$1.execute(JavaPluginLoader.java:312)
[06:41:26] [Server thread/WARN]:        at org.bukkit.plugin.RegisteredListener.callEvent(RegisteredListener.java:66)
[06:41:26] [Server thread/WARN]:        at org.bukkit.plugin.SimplePluginManager.fireEvent(SimplePluginManager.java:507)
[06:41:26] [Server thread/WARN]:        at org.bukkit.plugin.SimplePluginManager.callEvent(SimplePluginManager.java:492)
[06:41:26] [Server thread/WARN]:        at org.bukkit.craftbukkit.v1_13_R2.event.CraftEventFactory.callPlayerInteractEvent(CraftEventFactory.java:304)
[06:41:26] [Server thread/WARN]:        at net.minecraft.server.v1_13_R2.PlayerInteractManager.a(PlayerInteractManager.java:458)
[06:41:26] [Server thread/WARN]:        at net.minecraft.server.v1_13_R2.PlayerConnection.a(PlayerConnection.java:1222)
[06:41:26] [Server thread/WARN]:        at net.minecraft.server.v1_13_R2.PacketPlayInUseItem.a(PacketPlayInUseItem.java:37)
[06:41:26] [Server thread/WARN]:        at net.minecraft.server.v1_13_R2.PacketPlayInUseItem.a(PacketPlayInUseItem.java:1)
[06:41:26] [Server thread/WARN]:        at net.minecraft.server.v1_13_R2.PlayerConnectionUtils.lambda$0(PlayerConnectionUtils.java:9)
[06:41:26] [Server thread/WARN]:        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
[06:41:26] [Server thread/WARN]:        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
[06:41:26] [Server thread/WARN]:        at net.minecraft.server.v1_13_R2.SystemUtils.a(SourceFile:199)
[06:41:26] [Server thread/WARN]:        at net.minecraft.server.v1_13_R2.MinecraftServer.b(MinecraftServer.java:896)
[06:41:26] [Server thread/WARN]:        at net.minecraft.server.v1_13_R2.DedicatedServer.b(DedicatedServer.java:417)
[06:41:26] [Server thread/WARN]:        at net.minecraft.server.v1_13_R2.MinecraftServer.a(MinecraftServer.java:831)
[06:41:26] [Server thread/WARN]:        at net.minecraft.server.v1_13_R2.MinecraftServer.run(MinecraftServer.java:729)
[06:41:26] [Server thread/WARN]:        at java.lang.Thread.run(Thread.java:748)
```

**Comment by gorogoro-space** [comment]
Opps!
```
198|OAK_WALL_SIGN
```
I'm sorry. Garbage left when using 1.14.4. . . I will try to erase.

**Comment by gorogoro-space** [comment]
Doing this solved the error.
but, some of the protection data was broken. . .
I will try to recover from a database backup.
Thank you.

```
sqlite> delete from lwc_blocks where name = 'OAK_WALL_SIGN';
sqlite> .quit
```

**Comment by pop4959** [comment]
Ah yes, that makes sense. It would probably fail to load newer materials due to them not existing in older versions. I think I will still label this as a bug since loading the next index separately from the materials will probably make this more robust to this kind of situation. Glad you were able to figure it out on your own though! :)

**Comment by pop4959** [comment]
Should be fixed in 1350a90ac0702e98cb96438d53211e62d39f6637. LWC will still display a message if it isn't able to load a block (ie. block doesn't exist in the given version), but this will not cause the plugin to fail adding new blocks. The ID is determined separately now, so even with invalid blocks in the table, there should never be a unique constraint violation. :smile:
